Seg Array Optimization

CPBridge opened the pull request:
A few improvements focused on improving the efficiency of building segmentation arrays from loaded Segmentation objects.

This was achieved through:
- Re-ordering loops in the construction of the seg_frames_matrix to avoid searching through the large `_frames_lut` more than necessary.
- Creating a specialized case for the `combine_segments` option, rather than implementing it as a postprocessing of the standard routine. This reduces memory consumption significantly as it is now O(1) rather than O(n) in the number of segments requested.
- Being more careful with dtypes, and allowing the user to specify a dtype
- Adding an option to skip checks for overlapping segments (`skip_overlap_checks`), which results in a further modest speed-up.

@pieper you should now find that your example from #202 runs significantly faster and uses much less memory than before when `combine_segments=True` (especially in the case where the full set of segments is requested).

I went a little way down the path of JITing the methods with numba, but it actually slowed things down. There may be a bit more performance improvement possibly using threading, but we are hitting on some numpy limitations

pieper commented:
Thanks @CPBridge !  Let me know if you need me to test before you merge or i can just wait until @hackermd has had a look through it.

CPBridge commented:
Thanks Steve. If you do have time to test on your hardware at some point during the process that would be very helpful. I think before or after @hackermd has a look would be fine, I don't expect the core numerical process change

CPBridge commented:
A couple of other points to note:

- It will still be slow to load the file initially, that will require a fix in pydicom.
- Please try with and without the `skip_overlap_checks` option and see how this affects things

hackermd commented:
Thanks @CPBridge. I will take a look.

> It will still be slow to load the file initially, that will require a fix in pydicom

See https://github.com/pydicom/pydicom/pull/1734

hackermd commented on src/highdicom/seg/sop.py:
```diff
@@ -1916,81 +1984,140 @@ def _get_pixels_by_seg_frame(
+                                self.pixel_array
+                        else:
+                            out_array[out_frm, :, :, out_seg] = \
+                                self.pixel_array[seg_frame_ind, :, :]
```
Does this repeatedly decode the pixel data? It may be advantageous to store `pixel_array` as a separate variable outside of the loop instead of repeatedly accessing the attribute/property on the `self` object.

CPBridge replied:
No, this is cached on the Dataset at the pydicom level

hackermd replied:
Good to know. I would nevertheless suggest avoiding the repeated attribute access, since it may negatively impact performance. See https://github.com/pydicom/pydicom/issues/1728#issuecomment-1312833040

hackermd replied:
In context of the above comment (https://github.com/herrmannlab/highdicom/pull/208#discussion_r1042309582), we may even want to create a deep copy outside of the loop.

CPBridge replied:
A deepcopy out of the loop would be hugely inefficient and the comment you point to doesn't apply here since the `.pixel_array` exists as a Python-level property rather than the loopkup being deferred to pydicom.Dataset's dictionary lookup process

hackermd commented on src/highdicom/seg/sop.py:
```diff
@@ -2303,49 +2444,59 @@ def get_pixels_by_source_instance(
+                    f'SOP Instance UID {sop_uid} does not match any '
+                    'referenced source frames. To return an empty '
+                    'segmentation mask in this situation, use the '
+                    "'assert_missing_frames_are_empty' parameter."
```
For consistency

hackermd commented on src/highdicom/seg/sop.py:
```diff
@@ -2303,49 +2444,59 @@ def get_pixels_by_source_instance(
+        # Loop through each row of the LUT and determine whether the
+        # corresponding segmentation frame is required in the output, and if so,
+        # where
+        for seg_frm, lut_row in enumerate(lut):
+            # Source frame index and segment number for this LUT row
+            lut_src, lut_segno = lut_row.ravel()
+
+            # Find indices (if any) of these value in the lists of required
+            # output source frames and segments
+            out_frm_inds = np.where(frame_indices == lut_src)[0]
+            out_seg_inds = np.where(segment_numbers == lut_segno)[0]
```
I need to take a closer look, but at first glance I am wondering whether this is the most performant approach..

CPBridge replied:
See https://github.com/herrmannlab/highdicom/pull/208#issuecomment-1358495069

hackermd commented on src/highdicom/seg/sop.py:
```diff
@@ -2572,42 +2737,52 @@ def get_pixels_by_source_frame(
-                # else no segmentation frame found for this segment number,
-                # assume this frame is empty and leave the entry in seg_frames
-                # as -1
+        source_frame_numbers_arr = np.array(source_frame_numbers, np.uint)
+
+        # Loop through each row of the LUT and determine whether the
+        # corresponding segmentation frame is required in the output, and if so,
+        # where
+        for seg_frm, lut_row in enumerate(lut):
+            # Source frame index and segment number for this LUT row
+            lut_src, lut_segno = lut_row.ravel()
+
+            # Find indices (if any) of these value in the lists of required
+            # output source frames and segments
+            out_frm_inds = np.where(source_frame_numbers_arr == lut_src)[0]
+            out_seg_inds = np.where(segment_numbers == lut_segno)[0]
+
+            if len(out_frm_inds) > 1:
+                # This should never happen due to earlier checks
+                # on unique rows
+                raise RuntimeError(
+                    'An unexpected error was encountered during '
+                    'indexing of the segmentation image. Please '
+                    'file a bug report with the highdicom repository.'
+                )
+
+            # Frame is only required if there is an index for both
+            if len(out_frm_inds) > 0 and len(out_seg_inds) > 0:
+                r = out_frm_inds[0]
+                c = out_seg_inds[0]
+                seg_frames[r, c] = seg_frm + 1
```
This appears to be a duplication. Can we factor this out into a function?

CPBridge replied:
See https://github.com/herrmannlab/highdicom/pull/208#issuecomment-1358495069

hackermd reviewed (changes requested):
A few suggestions regarding style.

CPBridge commented on src/highdicom/seg/sop.py:
```diff
@@ -1876,6 +1932,18 @@ def _get_pixels_by_seg_frame(
+            overlap. However, this reduces performance. If checks are skipped
+            and multiple segments do overlap, the segment with the highest
+            segment number will be placed into the output array.
+        dtype: Union[type, None]
```
Resolved in 4325153

CPBridge commented on src/highdicom/seg/sop.py:
```diff
@@ -61,6 +61,60 @@
+    return dtype
+
+
+def _check_dtype(max_val: int, dtype: type) -> None:
```
Renamed to `_check_numpy_value_representation` in f80fd92, since "value representation" on its own may cause confusion with DICOM terminology

CPBridge commented on src/highdicom/seg/sop.py:
```diff
@@ -1916,81 +1984,140 @@ def _get_pixels_by_seg_frame(
+                        # Copy data to to output array
+                        if self.pixel_array.ndim == 2:
+                            # Special case with a single segmentation frame
+                            out_array[out_frm, :, :, out_seg] = \
+                                self.pixel_array
```
Can you elaborate on your concern, not sure I understand?

hackermd replied:
My concern is that NumPy may do a shallow copy and that `out_array` may continue to point to the array stored at `self.pixel_array` when it is returned by the function. With the references persisting, Python may not free the memory of `self.pixel_array` (even if the instance of the class is no longer needed).

It can sometimes be beneficial in my experience to enforce a deep copy for arrays that are returned from functions/methods.

https://numpy.org/doc/stable/user/quickstart.html#view-or-shallow-copy

CPBridge replied:
I'm almost certain that his isn't true after playing around with np.shared_memory but I added a .copy() just to be safe. Although the slicing operation could create a shallow copy, I'm pretty sure when you then assign this shallow copy as a value to a slice of the output array, the data is deepcopied. Otherwise the output array could consist of a bunch of different slices of other arrays, which is not allowed

CPBridge commented:
Ok @hackermd @pieper I completely refactored the way the lookup table is stored and queried and the result is much more efficient and in my opinion much less confusing than current master and even the previously reviewed version on this branch (which was itself much more efficient than master).

Previously I had implemented a lookup table as a numpy array and then wrote some frankly quite hacky code to "query" the lookup table to find the information I needed. I realised at the time that I was essentially implementing a relational database in numpy and that this was not a good idea, but I didn't want to introduce a new dependency...

...but at some point since then I became aware that a basic sqlite client is in the Python standard library and that this could deal with in-memory databases. I have refactored the entire lookup operation to use the standard library sqlite to construct and store the lookup table. Using sql syntax is much more ergonomic and in certain situations much more efficient than my nasty numpy implementation. In particular small queries are now very very fast. And all with no new dependencies. I'm just sad I didn't think of this before.

We should also consider this sort of approach elsewhere, e.g. storing a LUT of elements of an SR.

Putting this in the context of reducing overall seg parsing time, this has dramatically improved the efficiency of the "pixel reconstruction step". I'm now getting times of < 1s on my macbook for the full set of frames and segments on @pieper 's segmentation from #202 if I use `skip_overlap_checks`. (If someone can think of a way to make those checks more efficient I would love some ideas...). I believe there are still some gains to be had in the `_build_luts` method to extract the information to place into the SQL database from the PerFramesFunctionalGroupSequence (actually entering the information into the SQL database is very fast). But this PR is getting big and I think that is a separate thread that we can leave for the future.

hackermd commented on src/highdicom/seg/sop.py:
```diff
@@ -1328,35 +1387,51 @@ def from_dataset(
-
-        self._source_sop_instance_uids = np.array(
-            list(self._ref_ins_lut.keys())
+        cur.executemany(
+            "INSERT INTO InstanceUIDs "
+            "(StudyInstanceUID, SeriesInstanceUID, SOPInstanceUID) "
+            "VALUES(?, ?, ?)",
+            instance_data,
         )
+        self._db.commit()
```
You may want to consider managing transactions using [context manager](https://docs.python.org/3/library/sqlite3.html#sqlite3-connection-context-manager).

hackermd replied:
The same would apply to several other sections of the code.

CPBridge replied:
Nice, I didn't know about this. Implemented in b7f232f

hackermd commented on src/highdicom/seg/sop.py:
```diff
@@ -1372,11 +1447,12 @@ def _build_luts(self) -> None:
         index values.
 
         """
+        self._db = sqlite3.connect(":memory:")
```
You may need to be careful with the type of the `_db` attribute. We currently don't rely on mypy to check type annotations, but this would probably result in an error. See https://github.com/ImagingDataCommons/dicomweb-client/blob/3f91d8d8059a9f8fe117b1c28581ea814660478e/src/dicomweb_client/file.py#L560-L561

CPBridge replied:
Not really sure what the specific concern is here. The linked code appears to need type annotations because the connection variable may be `None` and may change type during the course of the programme to `sqlite3.Connection`. However that is not the case here: the `self._db` is never `None` so the annotation is not needed. I also checked with mypy and it doesn't raise any errors related to this.

If your concern is that the name of the variable is ambiguous, I changed it to `_db_con` in 2d1752e, and also added an explicit type annotation.

CPBridge replied:
(I accidentally changed to `db_con` instead of `_db_con` 2d1752e and corrected in b7f232f)

CPBridge commented on src/highdicom/seg/sop.py:
```diff
@@ -1942,7 +2044,7 @@ def _get_pixels_by_seg_frame(
             and multiple segments do overlap, the segment with the highest
             segment number will be placed into the output array.
-        dtype: Union[type, str, np.dtype, None]
+        dtype: Union[I
```
Oops, thanks for catching. Resolved in 6a647f9

CPBridge commented on src/highdicom/seg/sop.py:
```diff
@@ -1386,35 +1387,51 @@ def from_dataset(
+        cur.execute(
+            """
+                CREATE TABLE InstanceUIDs(
+                    SOPInstanceUID_id INTEGER PRIMARY KEY AUTOINCREMENT,
```
No it's not, but while prototyping I had it like this as I thought that using a pretty long string as a primary key (and more importantly as a foreign key in joins) would be quite inefficient compared to using an integer. But I just tested it and it seems to make no discernible difference, and it probably does make things more confusing, so I removed it in [f81e7b6](https://github.com/ImagingDataCommons/highdicom/pull/208/commits/f81e7b6b3020bd31b234049abbfd578bd67dea7c)

CPBridge commented:
@hackermd thanks for your comments, I have made some updates.

I also snuck in another change in [8e95876](https://github.com/ImagingDataCommons/highdicom/pull/208/commits/8e95876f85b2494f4845457f4704a719376cd22e) to change behaviour that annoyed me recently.

The code to extract a segmentation mask for a list of frames is very conservative: if you give it a frame number that is not referenced by the segmentation it will by default error out. This makes sense when thinking about SOPInstanceUIDs of single frame referenced images. But in the multiframe referenced image case, I think it reasonable to assume that if a requested frame number is not referenced in the segmentation image, but a frame with a frame number higher than the requested frame is referenced, the requested frame exists in the referenced image and the corresponding 2D mask should be returned as zeros.

hackermd commented on src/highdicom/seg/sop.py:
```diff
@@ -1823,38 +1920,118 @@ def segmented_property_types(self) -> List[CodedConcept]:
+        Raises
+        ------
+        KeyError:
+            If the input sop_instance_uid is not in the database.
```
```suggestion
            If the input `sop_instance_uid` is not in the database.
```

hackermd commented on src/highdicom/seg/sop.py:
```diff
@@ -1823,38 +1920,118 @@ def segmented_property_types(self) -> List[CodedConcept]:
-            )
-        return ind.item()
+    def _check_sop_uid_exists(self, sop_instance_uid: str) -> None:
+        """Checks whether a SOP Instance UID is in the database.
```
```suggestion
        """Check whether a SOP Instance UID is in the database.
```

hackermd commented on src/highdicom/seg/sop.py:
```diff
@@ -1328,35 +1387,50 @@ def from_dataset(
-        self._source_sop_instance_uids = np.array(
-            list(self._ref_ins_lut.keys())
-        )
+        with self._db_con:
```
I wonder whether it would make sense to abstract the database out of the `SOPClass` into a separate (private) class. That class could handle the database connection and expose methods to insert, find, and retrieve information. I feel that would be cleaner and would facilitate testing.

hackermd replied:
Such a database service class could also become useful for managing frame-level data access for other multi-frame SOP class instances in the future (Parametric Map, etc.).

CPBridge replied:
As of [8f5172f](https://github.com/ImagingDataCommons/highdicom/pull/208/commits/8f5172f2f692937eee71574ab81c4bc95e32d87e) the database related logic has been significantly refactored into a new `_SegDBManager` class

hackermd commented on src/highdicom/seg/sop.py:
```diff
@@ -1823,38 +1920,118 @@ def segmented_property_types(self) -> List[CodedConcept]:
+    ) -> None:
+        """Create a temporary table of desired segment in the SQLite DB.
+
+        Assumes the self._db_con object is currently being used as a context
```
That seems to be quite an assumption and asking for troubling..

CPBridge replied:
This no longer applies after all the of the recent refactoring. All methods on the new `SegDBManager` class handle their own commits to the database

hackermd reviewed (commented):
Looks good to me overall. My only suggestion would be to factor out the database management logic from the SOP Class into a separate helper/service class.

CPBridge commented:
@hackermd I have refactored out all the database logic and I think this is ready to go

hackermd commented on src/highdicom/seg/sop.py:
```diff
@@ -61,6 +75,723 @@
+        }
+
+    @contextmanager
+    def temp_table(
```
Should this method and the following be private methods (i.e., starting with `_`)?

I would also consider renaming them `_generate_temp_table`. I generally prefer function names to start with a verb.

hackermd commented on src/highdicom/seg/sop.py:
```diff
@@ -61,6 +75,723 @@
+            string per column.
+        column_data: Iterable[Sequence[Any]]
+            Column data to place into the table.
+
```
Documentation of `Return` or `Yield` missing.

hackermd commented on src/highdicom/seg/sop.py:
```diff
@@ -61,6 +75,723 @@
+            self._db_con.execute('DROP TABLE TemporarySegmentNumbers')
+
+    @contextmanager
+    def get_indices_by_source_instance(
```
```suggestion
    def iter_indices_by_source_instance(
```

hackermd commented on src/highdicom/seg/sop.py:
```diff
@@ -61,6 +75,723 @@
+        combine_segments: bool = False,
+        relabel: bool = False,
+    ) -> Generator[Iterator[Tuple[int, int, int]], None, None]:
+        """Get frame indices to create a segmentation mask for given instances.
```
```suggestion
        """Iterate over segmentation frame indices for given source image instances.
```

hackermd commented on src/highdicom/seg/sop.py:
```diff
@@ -61,6 +75,723 @@
+        Parameters
+        ----------
+        source_sop_instance_uids: str
+            SOP Instance UID of the source instances to for which segmentations
+            are requested.
```
```suggestion
            SOP Instance UID of the source instances for which segmentation
            image frames are requested.
```

hackermd commented on src/highdicom/seg/sop.py:
```diff
@@ -61,6 +75,723 @@
+            SOP Instance UID of the source instances to for which segmentations
+            are requested.
+        segment_numbers: Sequence[int]
+            Sequence containing segment numbers to include.
```
I would avoid a description of the type. That information is already provided via the type declaration.

```suggestion
            Numbers of segments to include.
```

hackermd commented on src/highdicom/seg/sop.py:
```diff
@@ -61,6 +75,723 @@
+                yield self._db_con.execute(query)
+
+    @contextmanager
+    def get_indices_by_source_frame(
```
```suggestion
    def iter_indices_by_source_frame(
```

hackermd replied:
It may be helpful to describe somewhere why the two different methods are needed. I am assuming that the `iter_indices_by_source_instance()` is needed if a multi-frame segmentation image references multiple single-frame image instances and the `iter_indices_by_source_frame()` is needed if a multi-frame segmentation image references multiple frames of a single multi-frame image instance. Is that understanding correct?

hackermd commented on src/highdicom/seg/sop.py:
```diff
@@ -61,6 +75,723 @@
+        combine_segments: bool = False,
+        relabel: bool = False,
+    ) -> Generator[Iterator[Tuple[int, int, int]], None, None]:
+        """Get frame indices to create a segmentation mask for given frames.
```
```suggestion
        """Iterate over frame indices for given source image frames.
```

hackermd commented on src/highdicom/seg/sop.py:
```diff
@@ -61,6 +75,723 @@
+                yield self._db_con.execute(query)
+
+    @contextmanager
+    def get_indices_by_dimension_index_values(
```
```suggestion
    def iter_indices_by_dimension_index_values(
```

hackermd commented on src/highdicom/seg/sop.py:
```diff
@@ -61,6 +75,723 @@
+        combine_segments: bool = False,
+        relabel: bool = False,
+    ) -> Generator[Iterator[Tuple[int, int, int]], None, None]:
+        """Get frame indices to create a segmentation mask for given frames.
```
```suggestion
        """Iterate over frame indices for given dimension index values.
```

hackermd commented on src/highdicom/seg/sop.py:
```diff
@@ -2228,6 +2982,19 @@ def get_pixels_by_source_instance(
+            and multiple segments do overlap, the segment with the highest
+            segment number (after relabelling, if applicable) will be placed
+            into the output array.
+        dtype: Union[type, str, np.dtype, None]
```
```suggestion
        dtype: Union[type, str, numpy.dtype, None]
```

hackermd commented on src/highdicom/seg/sop.py:
```diff
@@ -2285,68 +3052,46 @@ def get_pixels_by_source_instance(
-        )
+        ) as indices:
+
+            output_array = self._get_pixels_by_seg_frame(
+                num_output_frames=len(source_sop_instance_uids),
+                indices_iterator=indices,
+                segment_numbers=np.array(segment_numbers),
+                combine_segments=combine_segments,
+                relabel=relabel,
+                rescale_fractional=rescale_fractional,
+                skip_overlap_checks=skip_overlap_checks,
+                dtype=dtype,
+            )
+
+        return output_array
 
```
```suggestion
            return self._get_pixels_by_seg_frame(
                num_output_frames=len(source_sop_instance_uids),
                indices_iterator=indices,
                segment_numbers=np.array(segment_numbers),
                combine_segments=combine_segments,
                relabel=relabel,
                rescale_fractional=rescale_fractional,
                skip_overlap_checks=skip_overlap_checks,
                dtype=dtype,
            )

```

hackermd commented on src/highdicom/seg/sop.py:
```diff
@@ -2443,19 +3190,32 @@ def get_pixels_by_source_frame(
+            and multiple segments do overlap, the segment with the highest
+            segment number (after relabelling, if applicable) will be placed
+            into the output array.
+        dtype: Union[type, str, np.dtype, None]
```
```suggestion
        dtype: Union[type, str, numpy.dtype, None]
```

hackermd commented on src/highdicom/seg/sop.py:
```diff
@@ -2715,6 +3456,19 @@ def get_pixels_by_dimension_index_values(
+            and multiple segments do overlap, the segment with the highest
+            segment number (after relabelling, if applicable) will be placed
+            into the output array.
+        dtype: Union[type, str, np.dtype, None]
```
```suggestion
        dtype: Union[type, str, numpy.dtype, None]
```

CPBridge reviewed (commented):
@hackermd I have addressed the various comments, mostly just accepting all suggestions. I have attempted to improve the docstrings for the `_SegDBManager` class
